src/Storages/ObjectStorage/StorageObjectStorageSource.cpp

```diff
@@ -0,0 +1,624 @@
+
+    if (object_info)
+    {
+        LOG_TEST(logger, "Next key: {}", object_info->relative_path);
+    }
+
+    return object_info;
+}
+
+StorageObjectStorageSource::GlobIterator::GlobIterator(
+    ObjectStoragePtr object_storage_,
+    ConfigurationPtr configuration_,
+    const ActionsDAG::Node * predicate,
+    const NamesAndTypesList & virtual_columns_,
+    ContextPtr context_,
+    ObjectInfos * read_keys_,
+    size_t list_object_keys_size,
+    bool throw_on_zero_files_match_,
+    std::function<void(FileProgress)> file_progress_callback_)
+    : IIterator("GlobIterator")
+    , WithContext(context_)
+    , object_storage(object_storage_)
+    , configuration(configuration_)
+    , virtual_columns(virtual_columns_)
+    , throw_on_zero_files_match(throw_on_zero_files_match_)
+    , read_keys(read_keys_)
+    , file_progress_callback(file_progress_callback_)
+{
+    if (configuration->isNamespaceWithGlobs())
+    {
+        throw Exception(ErrorCodes::BAD_ARGUMENTS, "Expression can not have wildcards inside namespace name");
+    }
+    else if (configuration->isPathWithGlobs())
+    {
+        const auto key_with_globs = configuration_->getPath();
+        const auto key_prefix = configuration->getPathWithoutGlobs();
+        object_storage_iterator = object_storage->iterate(key_prefix, list_object_keys_size);
+
+        matcher = std::make_unique<re2::RE2>(makeRegexpPatternFromGlobs(key_with_globs));
+        if (!matcher->ok())
+        {
+            throw Exception(
+                ErrorCodes::CANNOT_COMPILE_REGEXP,
+                "Cannot compile regex from glob ({}): {}", key_with_globs, matcher->error());
+        }
+
+        recursive = key_with_globs == "/**";
+        filter_dag = VirtualColumnUtils::createPathAndFileFilterDAG(predicate, virtual_columns);
+    }
+    else
+    {
+        const auto object_key = configuration_->getPath();
+        auto object_metadata = object_storage->getObjectMetadata(object_key);
+        auto object_info = std::make_shared<ObjectInfo>(object_key, object_metadata);
+
+        object_infos.emplace_back(object_info);
+        if (read_keys)
+            read_keys->emplace_back(object_info);
+
+        if (file_progress_callback)
+            file_progress_callback(FileProgress(0, object_metadata.size_bytes));
+
+        is_finished = true;
+    }
+}
+
+StorageObjectStorage::ObjectInfoPtr StorageObjectStorageSource::GlobIterator::nextImpl(size_t processor)
+{
+    std::lock_guard lock(next_mutex);
+    auto object_info = nextImplUnlocked(processor);
+    if (object_info)
+    {
+        if (first_iteration)
+            first_iteration = false;
+    }
+    else if (first_iteration && throw_on_zero_files_match)
+    {
+        throw Exception(ErrorCodes::FILE_DOESNT_EXIST, "Can not match any files");
+    }
+    return object_info;
+}
+
+StorageObjectStorage::ObjectInfoPtr StorageObjectStorageSource::GlobIterator::nextImplUnlocked(size_t /* processor */)
```

Comment:
Why do we need `processor` argument in all `next` methods in iterators? I don't see it's used anywhere

Reply:
It is used here https://github.com/ClickHouse/ClickHouse/blob/1601115ff099beb0af46136e2500c351b54a0e48/src/Storages/S3Queue/S3QueueSource.cpp#L62
It is not new to this PR, was added several months ago. But it will be removed soon, I will make a PR soon which will remove it.